blocks/rich-text/index.js

```diff
@@ -558,6 +559,44 @@ export class RichText extends Component {
 		if ( keyCode === BACKSPACE ) {
 			this.onChange();
 		}
+
+		// `scrollToCaret` is called on `nodechange`, whereas calling it on
+		// `keyup` *when* moving to a new RichText element results in incorrect
+		// scrolling. Though the following allows false positives, it results
+		// in much smoother scrolling.
+		if ( keyCode !== BACKSPACE && keyCode !== ENTER ) {
+			this.scrollToCaret();
+		}
+	}
+
+	scrollToCaret() {
+		if ( ! this.props.isViewportSmall ) {
+			return;
+		}
+
+		const { top: caretTop } = this.getEditorSelectionRect();
+
+		const container = getScrollContainer( this.editor.getBody() );
+		if ( ! container ) {
+			return;
+		}
+
+		// When scrolling, avoid positioning the caret at the very top of
+		// the viewport, providing some "air" and some textual context for
+		// the user, and avoiding toolbars.
+		const graceOffset = 100;
+
+		// Avoid pointless scrolling by establishing a threshold under
+		// which scrolling should be skipped;
+		const epsilon = 10;
+		const delta = caretTop - graceOffset;
+
+		if ( Math.abs( delta ) > epsilon ) {
+			container.scrollTo(
```

Comment:
I'm finding this is not very durable, or at least I'm finding some browser inconsistencies. The `container` here will be `document.body`. If you open your browser's Developer Tools console right now (on GitHub.com) and enter:

```
document.body.scrollTo( 0, 500 );
```

I expect nothing will happen.

... That is, of course, unless you're running Safari or, more precisely, iOS Safari (though desktop works just as well), where it will scroll.

I'm not familiar with why this is different, though I suspect that when there is no explicit scroll container, Chrome and Firefox expect scrolls to be made via `window.scrollTo`.

Though the fact that `getScrollContainer` is returning `document.body` could itself be concerning since, as mentioned, it's not an explicit scroll container. It does qualify for `document.body.scrollHeight > document.body.clientHeight`. Maybe it's because there's an `overflow-x: hidden` ? I've had some previous experience with issues where even one direction of `overflow` has effects in both directions.

If we expect to call `scrollTo` on the result of `getScrollableContainer`, then I think we ought to normalize so that if the browser must call `window.scrollTo` on a non-explicit container, then `getScrollableContainer` should return `window`. This means either improving the detection of a scrollable container, or returning `window` when the result would otherwise be `document.body`.

Reply:
> I'm finding this is not very durable […] That is, of course, unless you're running Safari or, more precisely, iOS Safari (though desktop works just as well), where it will scroll.

Thanks for digging. The fact is I haven't tested this on Android yet, as I currently don't own any such device. 

> If we expect to call `scrollTo` on the result of `getScrollableContainer`, […]

All that sounds good, I'll see how much we can get out of the `body->window` normalization.

Reply:
Apologies for not testing this sooner — I only tested on iPhone and in the browser and it worked fine there.

I retested, and in the Android emulator it's working pretty damn well for me:

![android emulator](https://user-images.githubusercontent.com/1204802/38351621-9bf6ccf0-38b0-11e8-977f-51720f047d4d.gif)

I also tested using the Chrome "mobile simulator" (which is just desktop chrome with a small viewport and fancy mobile scrollbars, so it's not accurate), but I got a few JS errors there:

![browser](https://user-images.githubusercontent.com/1204802/38351639-b0d053d0-38b0-11e8-8eda-a5312104d3f8.gif)

Let me know how I can help with any other testing.